python/ray/train/xgboost/xgboost_trainer.py

```diff
@@ -1,15 +1,220 @@
-from typing import Any, Dict
+import logging
+from functools import partial
+from typing import Any, Dict, Optional
 
 import xgboost
 
+import ray.train
 from ray.train import Checkpoint
+from ray.train.constants import _DEPRECATED_VALUE, TRAIN_DATASET_KEY
 from ray.train.gbdt_trainer import GBDTTrainer
+from ray.train.trainer import GenDataset
 from ray.train.xgboost import RayTrainReportCallback
+from ray.train.xgboost.v2 import XGBoostTrainer as SimpleXGBoostTrainer
 from ray.util.annotations import PublicAPI
 
+logger = logging.getLogger(__name__)
 
+
+def _xgboost_train_fn_per_worker(
+    config: dict, label_column: str, num_boost_round: int, xgboost_train_kwargs: dict
+):
+    from xgboost.collective import CommunicatorContext
+
+    from ray.train._internal.session import get_session
+
+    checkpoint = ray.train.get_checkpoint()
+    starting_model = None
+    remaining_iters = num_boost_round
+    if checkpoint:
+        starting_model = RayTrainReportCallback.get_model(checkpoint)
+        starting_iter = starting_model.num_boosted_rounds()
+        remaining_iters = num_boost_round - starting_iter
+        logger.info(
+            f"Model loaded from checkpoint will train for "
+            f"additional {remaining_iters} iterations (trees) in order "
+            "to achieve the target number of iterations "
+            f"({num_boost_round=})."
+        )
+
+    train_dataset = ray.train.get_dataset_shard(TRAIN_DATASET_KEY)
+    train_df = train_dataset.to_pandas()
+
+    eval_datasets = {
+        k: d for k, d in get_session().dataset_shard.items() if k != TRAIN_DATASET_KEY
```

Comment:
Why use internal API here? Can we get eval datasets also with the `get_dataset_shard` API?

Reply:
You have to know the keys in advance to get them all with `get_dataset_shard` 😢  The user can specify any keys in `XGBoostTrainer(datasets)` so I have to access it privately this way.

Reply:
Ah got it. I am thinking if we can extract the key in the init function of `XGBoostTrainer`?

Reply:
Yeah, I can pass those keys through the partial args of this function.

Reply:
updated